### PR TITLE
use new stream decoder API

### PR DESCRIPTION
**Pull request by japaric**
fixes knurling-rs/defmt#575

marking as a draft because I want to update the manual snapshot tests before merging this

### CODE REVIEW

**Comment by jonas-schievink** [comment]
I think it would be good to set up the unit tests to always be built from source, that makes it easier to modify them

bors r+

**Comment by japaric** [comment]
> I think it would be good to set up the unit tests to always be built from source, that makes it easier to modify them

generally +1 to the idea of having the source in-tree. always rebuilding the snapshot binaries does have the complication that stuff can vary between systems and break the tests e.g. file paths, binary sizes, etc. resilience against that would require sort of wildcards in the .snap files (`flashing program ({digit*}.{digit*} KiB)`)) which I'm not sure insta supports?

File: src/main.rs [hunk 1]
```diff
@@ -325,16 +330,12 @@ fn decode_and_print_defmt_logs(
...
-                let num_bytes = buffer.len();
-                buffer.rotate_left(consumed);
-                buffer.truncate(num_bytes - consumed);
             }
 
             Err(defmt_decoder::DecodeError::UnexpectedEof) => break,
 
             Err(defmt_decoder::DecodeError::Malformed) => {
-                log::error!("failed to decode defmt data: {:x?}", buffer);
+                log::error!("failed to decode defmt data");
```
Comment (japaric) [inline src/main.rs:338]: by the way, @Urhengulas this looked like the right place to modify to implement #249 (I think)

Comment (Urhengulas) [reply src/main.rs:338]: Thank you for the pointer!

**Pull request merged**